File: api/app/domains/tax/tax_calc.py
"""Direct-tax computation core — pure, exact (integer paise), deterministic.

Covers the advance-tax schedule + s.234C deferment interest (with the 12%/36% relief
provisos), s.234E TDS-return late fee, the s.44AB tax-audit trigger, and MAT (s.115JB).
Rates/thresholds are **FY 2025-26 (AY 2026-27)** and declared as data — re-verify each
Finance Act (see skills/indian-fin-rules).
"""

from __future__ import annotations

from decimal import ROUND_HALF_UP, Decimal
from typing import Any

from app.core.money import Paise

# s.234C installments: (label, cumulative %, relief-floor %, months of interest).
# Q1/Q2 carry the statutory relief (no interest if >=12%/36% paid); Q3/Q4 do not.
_ADVANCE_TAX_SCHEDULE: list[tuple[str, Decimal, Decimal, int]] = [
    ("Q1", Decimal("0.15"), Decimal("0.12"), 3),
    ("Q2", Decimal("0.45"), Decimal("0.36"), 3),
    ("Q3", Decimal("0.75"), Decimal("0.75"), 3),
    ("Q4", Decimal("1.00"), Decimal("1.00"), 1),
]
_MONTHLY_INTEREST = Decimal("0.01")  # 1% per month (s.234B / 234C)

_234E_PER_DAY = Paise.from_rupees(200)

# s.44AB thresholds
_AUDIT_BUSINESS = Paise.from_rupees(10_000_000)  # ₹1 Cr
_AUDIT_BUSINESS_DIGITAL = Paise.from_rupees(100_000_000)  # ₹10 Cr (cash ≤ 5%)
_AUDIT_PROFESSIONAL = Paise.from_rupees(5_000_000)  # ₹50 L

_MAT_RATE = Decimal("0.15")  # s.115JB
_CESS = Decimal("0.04")


def _round_rupee(paise: Decimal) -> int:
    return int((paise / 100).quantize(Decimal("1"), rounding=ROUND_HALF_UP)) * 100


def advance_tax_schedule(total_liability: int) -> list[dict[str, Any]]:
    """Cumulative advance-tax target by installment."""
    return [
        {"installment": label, "cumulative_required": _round_rupee(Decimal(total_liability) * pct)}
        for label, pct, _floor, _m in _ADVANCE_TAX_SCHEDULE
    ]


def interest_234c(total_liability: int, cumulative_paid: list[int]) -> dict[str, Any]:
    """s.234C deferment interest given cumulative advance tax paid by each due date.
    ``cumulative_paid`` is 4 values (by 15 Jun/Sep/Dec/Mar). Relief: no interest for Q1/Q2
    if at least 12%/36% was paid."""
    if len(cumulative_paid) != 4:
        raise ValueError("cumulative_paid must have 4 entries (Q1..Q4)")
    total = 0
    per_installment: dict[str, int] = {}
    for (label, pct, floor_pct, months), paid in zip(
        _ADVANCE_TAX_SCHEDULE, cumulative_paid, strict=True
    ):
        required = Decimal(total_liability) * pct
        floor = Decimal(total_liability) * floor_pct
        if Decimal(paid) >= floor:
            per_installment[label] = 0
            continue
        shortfall = required - Decimal(paid)
        interest = _round_rupee(shortfall * _MONTHLY_INTEREST * months)
        per_installment[label] = interest
        total += interest
    return {"total_234c": total, "by_installment": per_installment}


def interest_234b(assessed_tax: int, advance_paid: int, *, months: int) -> dict[str, Any]:
    """s.234B interest: when advance tax paid is below 90% of assessed tax, 1%/month (simple)
    on the shortfall — the SHORTFALL (the interest base per s.234B(1)) rounded down to the
    nearest ₹100 (Rule 119A(c), Income-tax Rules 1962) — from 1 Apr of the assessment year to
    the date of payment (``months``)."""
    if assessed_tax <= 0 or months <= 0:
        return {"applicable": False, "shortfall": 0, "interest": 0, "months": months}
    if Decimal(advance_paid) >= Decimal(assessed_tax) * Decimal("0.9"):
        return {"applicable": False, "shortfall": 0, "interest": 0, "months": months}
    shortfall = max(0, assessed_tax - advance_paid)
    shortfall = (shortfall // 10000) * 10000  # round down to nearest ₹100 (10,000 paise)
    interest = _round_rupee(Decimal(shortfall) * _MONTHLY_INTEREST * months)
    return {"applicable": True, "shortfall": shortfall, "interest": interest, "months": months}


def late_fee_234e(days_late: int, tds_amount: int) -> int:
    """s.234E late fee for a TDS return: ₹200/day, capped at the TDS amount."""
    if days_late <= 0:
        return 0
    return min(int(_234E_PER_DAY) * int(days_late), int(tds_amount))


def tax_holiday_deduction(profit: int, *, claimed_years: int, eligible: bool) -> dict[str, Any]:
    """s.80-IAC: an eligible DPIIT-recognised startup may deduct 100% of profits for any 3
    consecutive AYs out of its first 10. ``claimed_years`` is how many of the 3 are already
    used. Grants the deduction this year only if eligible, profit is positive, and the 3-year
    allowance isn't exhausted."""
    available = eligible and profit > 0 and claimed_years < 3
    deduction = int(profit) if available else 0
    return {
        "eligible": available,
        "deduction": deduction,
        "taxable_after_holiday": int(profit) - deduction,
        "holiday_years_remaining": max(0, 3 - claimed_years - (1 if available else 0)),
    }


def reconcile_26as(books: list[dict], as_26as: list[dict]) -> dict[str, Any]:
    """Reconcile TDS credits in the books against Form 26AS (the department's record). Entries
    are aggregated by deductor TAN: {tan, amount}. Flags mismatches and one-sided entries."""
    book_by_tan: dict[str, int] = {}
    dept_by_tan: dict[str, int] = {}
    for e in books:
        book_by_tan[e["tan"]] = book_by_tan.get(e["tan"], 0) + int(e["amount"])
    for e in as_26as:
        dept_by_tan[e["tan"]] = dept_by_tan.get(e["tan"], 0) + int(e["amount"])

    matched: list[dict[str, Any]] = []
    mismatched: list[dict[str, Any]] = []
    missing_in_26as: list[dict[str, Any]] = []
    missing_in_books: list[dict[str, Any]] = []
    for tan in sorted(set(book_by_tan) | set(dept_by_tan)):
        bv, dv = book_by_tan.get(tan), dept_by_tan.get(tan)
        if dv is None:
            missing_in_26as.append({"tan": tan, "books": bv})
        elif bv is None:
            missing_in_books.append({"tan": tan, "as_26as": dv})
        elif bv == dv:
            matched.append({"tan": tan, "amount": bv})
        else:
            mismatched.append({"tan": tan, "books": bv, "as_26as": dv, "variance": bv - dv})
    return {
        "matched": matched,
        "mismatched": mismatched,
        "missing_in_26as": missing_in_26as,
        "missing_in_books": missing_in_books,
        "reconciled": not (mismatched or missing_in_26as or missing_in_books),
    }


def audit_required(
    turnover: int, *, cash_ratio: float = 0.0, is_professional: bool = False
) -> bool:
    """s.44AB tax-audit trigger. Business: > ₹10 Cr always; > ₹1 Cr if cash receipts/payments
    exceed 5%. Profession: gross receipts > ₹50 L."""
    if is_professional:
        return turnover > int(_AUDIT_PROFESSIONAL)
    if turnover > int(_AUDIT_BUSINESS_DIGITAL):
        return True
    return turnover > int(_AUDIT_BUSINESS) and cash_ratio > 0.05


# s.115BAA concessional regime (companies): 22% base + 10% surcharge + 4% cess = 25.168%
# effective (MMX-1.0 §WS1.C4). MAT (s.115JB) does NOT apply on this path.
_COMPANY_115BAA_BASE = Decimal("0.22")
_COMPANY_115BAA_SURCHARGE = Decimal("0.10")
_FIRM_TAX_RATE = Decimal("0.30")  # LLP / firm
_ONE_CRORE_PAISE = 10**7 * 100
_RULE_10D_THRESHOLD = _ONE_CRORE_PAISE  # Rs 1 crore aggregate international transactions
_MASTER_FILE_THRESHOLD = 500 * _ONE_CRORE_PAISE  # Rs 500 crore group revenue
_CBCR_THRESHOLD = 5500 * _ONE_CRORE_PAISE  # Rs 5500 crore group revenue (s.286)


def itr_computation(
    *,
    entity_type: str,
    gross_total_income: int,
    deductions: int = 0,
    book_profit: int | None = None,
    tds_paid: int = 0,
    advance_tax_paid: int = 0,
    regime_115baa: bool = True,
) -> dict[str, Any]:
    """Prepare the headline ITR computation. ``entity_type`` 'company' → ITR-6, 'firm'/'llp' →
    ITR-5 (30% + cess). The company regime is chosen explicitly (§WS1.C4): ``regime_115baa=True``
    (default) → 22% + 10% surcharge + 4% cess = 25.168% effective, and MAT (s.115JB) is EXCLUDED;
    the non-115BAA path (normal rates + MAT comparison) has no CA-initialled rate vector yet and is
    BLOCKED-CA. Prepaid TDS + advance tax are netted off. The e-filing upload is out of scope.
    Pure & exact paise."""
    et = entity_type.lower()
    total_income = max(0, int(gross_total_income) - int(deductions))
    if et == "company":
        if not regime_115baa:
            raise NotImplementedError(
                "BLOCKED-CA: non-115BAA company tax needs a CA-initialled rate vector "
                "(MMX-1.0 §0.6, §WS1.C4)."
            )
        form = "ITR-6"
        # 22% × 1.10 surcharge × 1.04 cess = 25.168% effective; MAT excluded on this path.
        effective = (
            _COMPANY_115BAA_BASE * (Decimal(1) + _COMPANY_115BAA_SURCHARGE) * (Decimal(1) + _CESS)
        )
        normal_tax = _round_rupee(Decimal(total_income) * effective)
        mat = 0
    else:
        form = "ITR-5"
        normal_tax = _round_rupee(Decimal(total_income) * _FIRM_TAX_RATE * (Decimal(1) + _CESS))
        mat = 0
    tax_payable = max(normal_tax, mat)
    prepaid = int(tds_paid) + int(advance_tax_paid)
    return {
        "form": form,
        "entity_type": et,
        "total_income": total_income,
        "normal_tax": normal_tax,
        "mat": mat,
        "tax_payable": tax_payable,
        "prepaid_taxes": prepaid,
        "balance_payable": max(0, tax_payable - prepaid),
        "refund_due": max(0, prepaid - tax_payable),
    }


def arms_length_check(
    price: int, comparables: list[int], *, tolerance_pct: float = 3.0
) -> dict[str, Any]:
    """Arm's-length test for a controlled transaction. The arm's-length price is the arithmetic
    mean of the uncontrolled comparables (Rule 10CA); a ±``tolerance_pct`` band (3% proviso)
    defines the acceptable range. Returns whether ``price`` is at arm's length and any TP
    adjustment to the ALP. Pure."""
    if not comparables:
        return {"at_arms_length": None, "reason": "no comparables provided"}
    mean = sum(int(c) for c in comparables) // len(comparables)
    band = int(
        (Decimal(mean) * Decimal(str(tolerance_pct)) / Decimal(100)).to_integral_value(
            ROUND_HALF_UP
        )
    )
    lower, upper = mean - band, mean + band
    at_arms_length = lower <= int(price) <= upper
    return {
        "at_arms_length": at_arms_length,
        "arms_length_price": mean,
        "lower": lower,
        "upper": upper,
        "adjustment": 0 if at_arms_length else mean - int(price),
    }


def tp_documentation_required(
    *, intl_transaction_value: int, group_consolidated_revenue: int = 0
) -> dict[str, Any]:
    """Transfer-pricing documentation thresholds: Form 3CEB whenever international transactions
    exist; Rule 10D documentation if their aggregate value exceeds ₹1 crore; Master File if group
    consolidated revenue exceeds ₹500 crore; CbCR (s.286) above ₹5,500 crore. Money in paise."""
    intl = int(intl_transaction_value)
    group = int(group_consolidated_revenue)
    return {
        "form_3ceb_required": intl > 0,
        "rule_10d_documentation": intl > _RULE_10D_THRESHOLD,
        "master_file_required": group > _MASTER_FILE_THRESHOLD,
        "cbcr_required": group > _CBCR_THRESHOLD,
    }


def mat_liability(book_profit: int) -> int:
    """Minimum Alternate Tax (s.115JB): 15% of book profit + 4% cess. Surcharge (which
    depends on income slab) is layered on by the caller when applicable."""
    if book_profit <= 0:
        return 0
    base = Decimal(book_profit) * _MAT_RATE
    return _round_rupee(base * (Decimal(1) + _CESS))

File: api-nest/src/domains/tax/tax.calc.ts
/**
 * Direct-tax computation core — pure, exact (integer paise), deterministic.
 * Faithful port of api/app/domains/tax/tax_calc.py.
 *
 * Covers advance-tax schedule + s.234C deferment interest (with the 12%/36% relief
 * provisos), s.234B, s.234E TDS-return late fee, s.44AB tax-audit trigger, MAT (s.115JB),
 * ITR headline, transfer pricing, 26AS reconciliation, s.80-IAC tax holiday.
 * Rates/thresholds are FY 2025-26 (AY 2026-27); re-verify each Finance Act.
 *
 * Money is integer paise throughout. Python `Decimal ... ROUND_HALF_UP` (half away from zero)
 * is replicated with exact BigInt arithmetic so there is zero float drift on money.
 */

// Round half away from zero for the rational a/b (b > 0). BigInt = exact.
function divRoundHalfAway(a: bigint, b: bigint): bigint {
  const neg = a < 0n;
  const aa = neg ? -a : a;
  const q = (2n * aa + b) / (2n * b); // floor(aa/b + 1/2) = round-half-up for non-negative
  return neg ? -q : q;
}

/**
 * Python `_round_rupee(paise)`: round the (possibly fractional) paise value paiseNum/paiseDen
 * to the nearest rupee half-away-from-zero, and return integer paise. Exact via BigInt.
 */
function roundRupeeExact(paiseNum: bigint, paiseDen: bigint): number {
  const rupees = divRoundHalfAway(paiseNum, paiseDen * 100n);
  return Number(rupees * 100n);
}

const B = (n: number): bigint => BigInt(Math.trunc(n));

// s.234C installments: cumulative % (×100), relief-floor % (×100), months of interest.
// Q1/Q2 carry the statutory relief (no interest if >=12%/36% paid); Q3/Q4 do not.
type Installment = { label: string; pctNum: number; floorNum: number; months: number };
const ADVANCE_TAX_SCHEDULE: Installment[] = [
  { label: 'Q1', pctNum: 15, floorNum: 12, months: 3 },
  { label: 'Q2', pctNum: 45, floorNum: 36, months: 3 },
  { label: 'Q3', pctNum: 75, floorNum: 75, months: 3 },
  { label: 'Q4', pctNum: 100, floorNum: 100, months: 1 },
];

const _234E_PER_DAY = 20000; // Paise.from_rupees(200)

// s.44AB thresholds (paise)
const AUDIT_BUSINESS = 1_000_000_000; // ₹1 Cr
const AUDIT_BUSINESS_DIGITAL = 10_000_000_000; // ₹10 Cr (cash ≤ 5%)
const AUDIT_PROFESSIONAL = 500_000_000; // ₹50 L

const ONE_CRORE_PAISE = 10 ** 7 * 100; // 1e9
const RULE_10D_THRESHOLD = ONE_CRORE_PAISE; // ₹1 Cr aggregate international transactions
const MASTER_FILE_THRESHOLD = 500 * ONE_CRORE_PAISE; // ₹500 Cr group revenue
const CBCR_THRESHOLD = 5500 * ONE_CRORE_PAISE; // ₹5500 Cr group revenue (s.286)

// ---- advance tax / s.234C -----------------------------------------------------

export function advanceTaxSchedule(totalLiability: number): Array<{ installment: string; cumulative_required: number }> {
  const liab = B(totalLiability);
  return ADVANCE_TAX_SCHEDULE.map(({ label, pctNum }) => ({
    installment: label,
    // paise = liability * pctNum / 100
    cumulative_required: roundRupeeExact(liab * BigInt(pctNum), 100n),
  }));
}

export function interest234c(
  totalLiability: number,
  cumulativePaid: number[],
): { total_234c: number; by_installment: Record<string, number> } {
  if (cumulativePaid.length !== 4) {
    throw new Error('cumulative_paid must have 4 entries (Q1..Q4)');
  }
  const liab = B(totalLiability);
  let total = 0;
  const byInstallment: Record<string, number> = {};
  ADVANCE_TAX_SCHEDULE.forEach(({ label, pctNum, floorNum, months }, i) => {
    const paid = B(cumulativePaid[i]);
    // relief: Decimal(paid) >= liability*floorNum/100  ⟺  100*paid >= liability*floorNum
    if (100n * paid >= liab * BigInt(floorNum)) {
      byInstallment[label] = 0;
      return;
    }
    // shortfall = liability*pctNum/100 - paid; interest = round_rupee(shortfall * 0.01 * months)
    // paise = (liability*pctNum - 100*paid) * months / (100 * 100)
    const paiseNum = (liab * BigInt(pctNum) - 100n * paid) * BigInt(months);
    const interest = roundRupeeExact(paiseNum, 10000n);
    byInstallment[label] = interest;
    total += interest;
  });
  return { total_234c: total, by_installment: byInstallment };
}

export function interest234b(
  assessedTax: number,
  advancePaid: number,
  months: number,
): { applicable: boolean; shortfall: number; interest: number; months: number } {
  if (assessedTax <= 0 || months <= 0) {
    return { applicable: false, shortfall: 0, interest: 0, months };
  }
  // advance_paid >= assessed * 0.9  ⟺  10*advance_paid >= 9*assessed
  if (10 * advancePaid >= 9 * assessedTax) {
    return { applicable: false, shortfall: 0, interest: 0, months };
  }
  let shortfall = Math.max(0, assessedTax - advancePaid);
  shortfall = Math.trunc(shortfall / 10000) * 10000; // round down to nearest ₹100 (10,000 paise)
  // interest = round_rupee(shortfall * 0.01 * months) => paise = shortfall*months/100
  const interest = roundRupeeExact(B(shortfall) * B(months), 100n);
  return { applicable: true, shortfall, interest, months };
}

// ---- s.234E TDS-return late fee -----------------------------------------------

export function lateFee234e(daysLate: number, tdsAmount: number): number {
  if (daysLate <= 0) return 0;
  return Math.min(_234E_PER_DAY * Math.trunc(daysLate), Math.trunc(tdsAmount));
}

// ---- s.80-IAC tax holiday -----------------------------------------------------

export function taxHolidayDeduction(
  profit: number,
  claimedYears: number,
  eligible: boolean,
): { eligible: boolean; deduction: number; taxable_after_holiday: number; holiday_years_remaining: number } {
  const available = eligible && profit > 0 && claimedYears < 3;
  const deduction = available ? Math.trunc(profit) : 0;
  return {
    eligible: available,
    deduction,
    taxable_after_holiday: Math.trunc(profit) - deduction,
    holiday_years_remaining: Math.max(0, 3 - claimedYears - (available ? 1 : 0)),
  };
}

// ---- 26AS reconciliation ------------------------------------------------------

type TanEntry = { tan: string; amount: number };

export function reconcile26as(
  books: TanEntry[],
  as26as: TanEntry[],
): {
  matched: Array<{ tan: string; amount: number }>;
  mismatched: Array<{ tan: string; books: number; as_26as: number; variance: number }>;
  missing_in_26as: Array<{ tan: string; books: number }>;
  missing_in_books: Array<{ tan: string; as_26as: number }>;
  reconciled: boolean;
} {
  const bookByTan = new Map<string, number>();
  const deptByTan = new Map<string, number>();
  for (const e of books) bookByTan.set(e.tan, (bookByTan.get(e.tan) ?? 0) + Math.trunc(e.amount));
  for (const e of as26as) deptByTan.set(e.tan, (deptByTan.get(e.tan) ?? 0) + Math.trunc(e.amount));

  const matched: Array<{ tan: string; amount: number }> = [];
  const mismatched: Array<{ tan: string; books: number; as_26as: number; variance: number }> = [];
  const missing_in_26as: Array<{ tan: string; books: number }> = [];
  const missing_in_books: Array<{ tan: string; as_26as: number }> = [];

  const tans = [...new Set([...bookByTan.keys(), ...deptByTan.keys()])].sort();
  for (const tan of tans) {
    const bv = bookByTan.get(tan);
    const dv = deptByTan.get(tan);
    if (dv === undefined) {
      missing_in_26as.push({ tan, books: bv as number });
    } else if (bv === undefined) {
      missing_in_books.push({ tan, as_26as: dv });
    } else if (bv === dv) {
      matched.push({ tan, amount: bv });
    } else {
      mismatched.push({ tan, books: bv, as_26as: dv, variance: bv - dv });
    }
  }
  return {
    matched,
    mismatched,
    missing_in_26as,
    missing_in_books,
    reconciled: mismatched.length === 0 && missing_in_26as.length === 0 && missing_in_books.length === 0,
  };
}

// ---- s.44AB tax-audit trigger -------------------------------------------------

export function auditRequired(
  turnover: number,
  opts: { cashRatio?: number; isProfessional?: boolean } = {},
): boolean {
  const cashRatio = opts.cashRatio ?? 0.0;
  if (opts.isProfessional) return turnover > AUDIT_PROFESSIONAL;
  if (turnover > AUDIT_BUSINESS_DIGITAL) return true;
  return turnover > AUDIT_BUSINESS && cashRatio > 0.05;
}

// ---- ITR headline computation -------------------------------------------------

const COMPANY_RATE_NUM = 2288; // 0.22 * 1.04 * 10000 (s.115BAA 22% + 4% cess)
const FIRM_RATE_NUM = 3120; // 0.30 * 1.04 * 10000 (LLP / firm)

export function itrComputation(args: {
  entityType: string;
  grossTotalIncome: number;
  deductions?: number;
  bookProfit?: number | null;
  tdsPaid?: number;
  advanceTaxPaid?: number;
}): {
  form: string;
  entity_type: string;
  total_income: number;
  normal_tax: number;
  mat: number;
  tax_payable: number;
  prepaid_taxes: number;
  balance_payable: number;
  refund_due: number;
} {
  const et = args.entityType.toLowerCase();
  const totalIncome = Math.max(0, Math.trunc(args.grossTotalIncome) - Math.trunc(args.deductions ?? 0));
  const isCompany = et === 'company';
  const form = isCompany ? 'ITR-6' : 'ITR-5';
  const rateNum = isCompany ? COMPANY_RATE_NUM : FIRM_RATE_NUM;
  // normal_tax = round_rupee(total_income * rate * 1.04) => paise = total_income*rateNum/10000
  const normalTax = roundRupeeExact(B(totalIncome) * BigInt(rateNum), 10000n);
  const bp = args.bookProfit;
  const mat = isCompany && bp !== null && bp !== undefined ? matLiability(Math.trunc(bp)) : 0;
  const taxPayable = Math.max(normalTax, mat);
  const prepaid = Math.trunc(args.tdsPaid ?? 0) + Math.trunc(args.advanceTaxPaid ?? 0);
  return {
    form,
    entity_type: et,
    total_income: totalIncome,
    normal_tax: normalTax,
    mat,
    tax_payable: taxPayable,
    prepaid_taxes: prepaid,
    balance_payable: Math.max(0, taxPayable - prepaid),
    refund_due: Math.max(0, prepaid - taxPayable),
  };
}

// ---- transfer pricing ---------------------------------------------------------

export function armsLengthCheck(
  price: number,
  comparables: number[],
  tolerancePct = 3.0,
):
  | { at_arms_length: null; reason: string }
  | { at_arms_length: boolean; arms_length_price: number; lower: number; upper: number; adjustment: number } {
  if (comparables.length === 0) {
    return { at_arms_length: null, reason: 'no comparables provided' };
  }
  const sum = comparables.reduce((s, c) => s + Math.trunc(c), 0);
  const mean = Math.floor(sum / comparables.length); // Python // (floor)
  // band = int(Decimal(mean) * tol / 100)  — truncate toward zero.
  // ponytail: float tol matches Python Decimal(str(tol)) for ordinary percentages; exotic
  //           tolerances with long decimals could drift — swap to a rational if that matters.
  const band = Math.trunc((mean * tolerancePct) / 100);
  const lower = mean - band;
  const upper = mean + band;
  const atArmsLength = lower <= Math.trunc(price) && Math.trunc(price) <= upper;
  return {
    at_arms_length: atArmsLength,
    arms_length_price: mean,
    lower,
    upper,
    adjustment: atArmsLength ? 0 : mean - Math.trunc(price),
  };
}

export function tpDocumentationRequired(args: {
  intlTransactionValue: number;
  groupConsolidatedRevenue?: number;
}): {
  form_3ceb_required: boolean;
  rule_10d_documentation: boolean;
  master_file_required: boolean;
  cbcr_required: boolean;
} {
  const intl = Math.trunc(args.intlTransactionValue);
  const group = Math.trunc(args.groupConsolidatedRevenue ?? 0);
  return {
    form_3ceb_required: intl > 0,
    rule_10d_documentation: intl > RULE_10D_THRESHOLD,
    master_file_required: group > MASTER_FILE_THRESHOLD,
    cbcr_required: group > CBCR_THRESHOLD,
  };
}

// ---- MAT (s.115JB) ------------------------------------------------------------

export function matLiability(bookProfit: number): number {
  if (bookProfit <= 0) return 0;
  // 15% of book profit + 4% cess => paise = book_profit * 1560 / 10000
  return roundRupeeExact(B(bookProfit) * 1560n, 10000n);
}
